test/test_prepared.py

```diff
@@ -0,0 +1,46 @@
+def test_prepared_statement(bouncer):
+    bouncer.admin(f"set pool_mode=transaction")
+    bouncer.admin(f"set prepared_statement_cache_size=100")
+    prepared_query = "SELECT 1"
+    with bouncer.cur() as cur1:
+        with bouncer.cur() as cur2:
+            cur1.execute(prepared_query, prepare=True)
+            cur1.execute(prepared_query)
+            with cur2.connection.transaction():
+                cur2.execute("SELECT 2")
+                cur1.execute(prepared_query)
+
+
+def test_prepared_statement_params(bouncer):
+    bouncer.admin(f"set pool_mode=transaction")
+    bouncer.admin(f"set prepared_statement_cache_size=100")
+    prepared_query = "SELECT %s"
+    with bouncer.cur() as cur1:
+        with bouncer.cur() as cur2:
+            cur1.execute(prepared_query, params=(1,), prepare=True)
+            cur1.execute(prepared_query, params=(1,))
+            with cur2.connection.transaction():
+                cur2.execute("SELECT 2")
+                cur1.execute(prepared_query, params=(1,))
+
+
+def test_parse_larger_than_pkt_buf(bouncer):
+    bouncer.admin(f"set pool_mode=transaction")
+    bouncer.admin(f"set prepared_statement_cache_size=100")
+    long_string = "1" * 100000
+    prepared_query = "SELECT '" + long_string + "'"
```

Comment:
I'm working on a fix for this problem btw. It requires some serious refactoring.

Reply:
> I'm working on a fix for this problem btw. It requires some serious refactoring.

Do I correctly understand that the problem is caused by  calling `pktbuf_send_queued` multiple times in prepared_statemen.c ? If so, why there are no assert failures  in `pktbuf_send_queued`:
```
bool pktbuf_send_queued(PktBuf *buf, PgSocket *sk)
{
	Assert(!buf->sending);
	Assert(!buf->fixed_buf);
       ...
```

Reply:
The problem is that pktbuf_send_qeueued doesn't actually queue it with respect to where Sbuf currently is in respect to processing the data. It might send data too early. Or as in this case it sends part of the data that the OS buffer currently accepts, and queues the rest for later. But some other packet received from the client can be forwarded earlier because it needs to be sent according to pktbuf_prepare_send. So far it was only used in admin connections used `pktbuf_send_queued` where this problem is MUCH less likely to occur (and also less of a practical problem when it does).

I'm now ammending Sbuf so that you can provide it with an MBuf that should be sent instead of (part of) the packet in its IObuf.

Reply:
I pushed an initial version of this. It sadly still doesn't seem to help for Windows or Ubuntu 18.04. FreeBSD seems to be fixed though. I'll try to apply the same approach in a bit more places tomorrow. Maybe that will resolve the other failures. If not we'll have to investigate a bit more.

Reply:
Okay I found the cause for the Ubuntu 18.04 and Windows failures and fixed it in this commit: [`84e994a` (#845)](https://github.com/pgbouncer/pgbouncer/pull/845/commits/84e994a4d7ca63f49949b761778abc23dd5cc1c4)

For those two the default client_encoding on the server was different than the one that the client would set. PgBouncer would notice this in `find_server` while handling the Parse packet. Then to fix it it would send a `SET` to the server and it would pause the client and fail the handle of the the Parse packet until it received a response to the SET. This meant that the Parse packet would need to be reprocessed when the client got unpaused, and that didn't happen because the callback flag was now set to `CB_HANDLE_COMPLETE_PACKET` and that flag had no handler.